### PR TITLE
Add webhook trigger

### PR DESCRIPTION
**Pull request by bardielle**
This PR adds a rest API for `wehbhook triggers`, allowing you to trigger a new `OSBuild` CR by sending a request to the OCP endpoint. 
The new rest API is a different pod that runs separately from the operator. 

For all webhooks, you must define a secret with a key named `WebHookSecretKey` and the value being the value to be supplied when invoking the webhook.
The request should includes the secret value in the request's header and looks like:` curl -v -H "secret: secretvalue" -X POST http://<SERVICE_IP>:8080/api/osbuild/v1/namespaces/<osbuildconfig_namespace>/osbuildconfig/<osbuildconfig_name>/webhooks`

### CODE REVIEW

File: controllers/osbuildconfig_controller.go [hunk 1]
```diff
@@ -18,16 +18,14 @@ package controllers
 
 import (
 	"context"
-	"github.com/go-logr/logr"
 	osbuilderprojectflottaiov1alpha1 "github.com/project-flotta/osbuild-operator/api/v1alpha1"
+	"github.com/project-flotta/osbuild-operator/internal/maniffests"
```
Comment (ygalblum) [inline controllers/osbuildconfig_controller.go:22]: If there a reason for the double `f` in `maniffests` or is it a typo?

File: internal/maniffests/osbuild.go [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+	"github.com/project-flotta/osbuild-operator/internal/repository/osbuildconfig"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+func CreateNewOSBuildCR(ctx context.Context, osBuildConfig *osbuilderprojectflottaiov1alpha1.OSBuildConfig,
+	logger logr.Logger, osBuildConfigRepository osbuildconfig.Repository, osBuildRepository osbuild.Repository,
+	scheme *runtime.Scheme) error {
+	osBuildNewVersion := *osBuildConfig.Status.LastVersion + 1
```
Comment (ygalblum) [inline internal/maniffests/osbuild.go:19]: I know you copied this code, but it just occurred to me that the first time you create an `OSBuild` CR, `LastVersion` will be `nil`. So, you need to check it

File: internal/osbuilconfig/osbuilconfig_handler.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
+package osbuilconfig
```
Comment (ygalblum) [inline internal/osbuilconfig/osbuilconfig_handler.go:1]: You're missing a `d` in `osbuilconfig`

File: internal/osbuilconfig/osbuilconfig_handler.go [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+	logger := log.FromContext(ctx)
+	logger.Info("OSBuildConfigWebhookTriggerHandler new request")
+
+	osBuildConfig, err := o.OSBuildConfigRepository.Read(ctx, params.OsbuildconfigCrName, params.OsbuildconfigNamespaceName)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return osbuilconfig.NewOSBuildConfigWebhookTriggersNotFound()
+		}
+		logger.Error(err, "can't read OSBuildConfigCR")
+		return osbuilconfig.NewOSBuildConfigWebhookTriggersBadRequest()
```
Comment (ygalblum) [inline internal/osbuilconfig/osbuilconfig_handler.go:39]: Why is this a BadRequest and not an InternalServerError?

File: main.go [hunk 5]
```diff
@@ -18,8 +18,10 @@ package main
 
 import (
 	"flag"
-	"github.com/project-flotta/osbuild-operator/internal/repository/osbuild"
-	"github.com/project-flotta/osbuild-operator/internal/repository/osbuildconfig"
+	"fmt"
+	"github.com/project-flotta/osbuild-operator/restapi"
```
Comment (ygalblum) [inline main.go:22]: I think the imports from inside the project should be at the end

File: internal/osbuilconfig/osbuilconfig_handler.go (hunk 4 above)
Comment (bardielle) [reply internal/osbuilconfig/osbuilconfig_handler.go]: I wasn't sure here. 
If you think it should be internalServerError I will change it

File: config/manager/kustomization.yaml [hunk 6]
```diff
@@ -5,6 +5,12 @@ generatorOptions:
   disableNameSuffixHash: true
 
 configMapGenerator:
-- name: manager-config
-  files:
+- files:
```
Comment (ygalblum) [inline config/manager/kustomization.yaml:8]: Revert this change

**Review by ygalblum — CHANGES_REQUESTED** [review]

File: swagger.yaml [hunk 7]
```diff
@@ -0,0 +1,81 @@
+swagger: '2.0'
```
Comment (masayag) [inline swagger.yaml:1]: New restful API should be written using [openapi v3 ](https://swagger.io/specification/)

File: swagger.yaml [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+      tags:
+        - osbuildconfig
+      parameters:
+        - in: path
+          name: osbuildconfig_namespace_name
+          description: OSBuilfConfig namespace name
+          type: string
+          required: true
+        - in: path
+          name: osbuildconfig_cr_name
```
Comment (masayag) [inline swagger.yaml:45]: just `name` .

e.g. pod HTTP API:

```
HTTP Request 
GET /api/v1/namespaces/{namespace}/pods/{name}
```

File: swagger.yaml [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+          description: OSBuildConfig name
+          type: string
+          required: true
+        - in: path
+          name: secret
+          description: secret name
```
Comment (masayag) [inline swagger.yaml:51]: what is the content of the secret? does it contain a sensitive information? should it be specified in the path?

Comment (ygalblum) [reply swagger.yaml:51]: Please see: https://docs.openshift.com/container-platform/4.7/cicd/builds/triggering-builds-build-hooks.html#builds-using-generic-webhooks_triggering-builds-build-hooks

File: Makefile [hunk 10]
```diff
@@ -52,6 +52,7 @@ IMG ?= controller:latest
 ENVTEST_K8S_VERSION = 1.23
 
 CERT_MANAGER_VERSION = v1.8.0
+PWD=$(shell pwd)
```
Comment (ygalblum) [inline Makefile:55]: Is this a leftover? I don't see where you're using it

File: internal/osbuildconfig/osbuildconfig_handler.go [hunk 11]
```diff
@@ -0,0 +1,78 @@
...
+
+func NewOSBuildConfigHandler(osBuildConfigRepository repositoryosbuildconfig.Repository, osBuildRepository repositoryosbuild.Repository, secretRepository secret.Repository, scheme *runtime.Scheme, osBuildCRCreator manifests.OSBuildCRCreatorInterface) *OSBuildConfigHandler {
+	return &OSBuildConfigHandler{
+		OSBuildConfigRepository: osBuildConfigRepository,
+		OSBuildRepository:       osBuildRepository,
+		SecretRepository:        secretRepository,
+		Scheme:                  scheme,
+		OSBuildCRCreator:        osBuildCRCreator,
+	}
+}
```
Comment (ygalblum) [inline internal/osbuildconfig/osbuildconfig_handler.go:30]: Did you run `make fmt`? There should be an empty line here

File: internal/osbuildconfig/osbuildconfig_handler.go [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+func (o *OSBuildConfigHandler) OSBuildConfigWebhookTriggers(w http.ResponseWriter, r *http.Request, osbuildconfigNamespaceName string, osbuildconfigCrName string, secret string) {
```
Comment (ygalblum) [inline internal/osbuildconfig/osbuildconfig_handler.go:31]: You resolved @masayag's comment. But, weren't the parameter names supposed to be shorted (i.e. namespace and name)?

File: Makefile (hunk 10 above)
Comment (bardielle) [reply Makefile:55]: yes

File: internal/osbuildconfig/osbuildconfig_handler.go (hunk 11 above)
Comment (bardielle) [reply internal/osbuildconfig/osbuildconfig_handler.go:30]: yes, it didn't change something

File: internal/osbuildconfig/osbuildconfig_handler.go (hunk 12 above)
Comment (bardielle) [reply internal/osbuildconfig/osbuildconfig_handler.go]: fixed it thanks

File: osbuildconfig_webhook_trigger.yaml [hunk 13]
```diff
@@ -0,0 +1,78 @@
...
+    url: https://github.com/project-flotta
+    email: flotta@redhat.com
+  license:
+    name: Apache 2.0
+    url: http://www.apache.org/licenses/LICENSE-2.0.html
+tags:
+  - name: osbuilconfig
+    description: OSBuildConfig CRD
+paths:
+  "/namespaces/{namespace}/osbuildconfig/{name}/webhooks/{secret}/generic":
```
Comment (masayag) [inline osbuildconfig_webhook_trigger.yaml:17]: what are the future plans for this API? is it expected to evolve? support multiple webhook types or will it remain as a single path?

not sure if secret should be provided in the path rather in the body.
also, what `generic` stands for?

I'd see it as
POST `/namespaces/{namespace}/osbuildconfig/{name}/triggers` with secret in requestBody.
should create a new trigger.

Comment (bardielle) [reply osbuildconfig_webhook_trigger.yaml:17]: take a look here: https://docs.openshift.com/container-platform/4.7/cicd/builds/triggering-builds-build-hooks.html#builds-using-generic-webhooks_triggering-builds-build-hooks
it's a reference for me, secret (value)  is part of the path
https://<openshift_api_host:port>/apis/build.openshift.io/v1/namespaces/<namespace>/buildconfigs/<name>/webhooks/<secret>/github

File: internal/osbuildconfig/osbuildconfig_handler.go [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+		if errors.IsNotFound(err) {
+			w.WriteHeader(http.StatusNotFound)
+			return
+		}
+		logger.Error(err, "can't read secret")
+		w.WriteHeader(http.StatusInternalServerError)
+		return
+	}
+
+	if string(webhookSecret.Data["WebHookSecretKey"]) != secretVal {
```
Comment (masayag) [inline internal/osbuildconfig/osbuildconfig_handler.go:66]: according to the API, the secret provided to this method is the name of the secret. it isn't the value as it is being referred to here.
is the expectation to have the content of a secret part of the path?
also, if there are assumptions about the secret, e.g. the requirement to have a key named `WebHookSecretKey` it should be well documented.

File: osbuildconfig_webhook_trigger.yaml (hunk 13 above)
Comment (bardielle) [reply osbuildconfig_webhook_trigger.yaml:17]: No, as far as I know

File: internal/osbuildconfig/osbuildconfig_handler.go (hunk 11 above)
Comment (ygalblum) [reply internal/osbuildconfig/osbuildconfig_handler.go:30]: Strange. Anyhow, there should be an empty line here

File: internal/osbuildconfig/osbuildconfig_handler.go (hunk 14 above)
Comment (bardielle) [reply internal/osbuildconfig/osbuildconfig_handler.go:66]: I changed it, it was a bug that I fix but forgot to update the API 
It's a secret value - please read the new description, is it more clear now?

Comment (ygalblum) [reply internal/osbuildconfig/osbuildconfig_handler.go:66]: No, the secret provided is the value of the key `WebHookSecretKey`. It is documented here: https://github.com/project-flotta/osbuild-operator/blob/main/config/crd/bases/osbuilder.project-flotta.io_osbuildconfigs.yaml#L179

File: internal/osbuildconfig/osbuildconfig_handler.go [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+	logger.Info("OSBuildConfigWebhookTriggerHandler new request")
+
+	// TODO - understand how to get params and use the responses
+	osBuildConfig, err := o.OSBuildConfigRepository.Read(r.Context(), name, namespace)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			w.WriteHeader(http.StatusNotFound)
+			return
+		}
+		logger.Error(err, "can't read OSBuildConfigCR")
```
[the remark this answers is not in the document]
Comment (bardielle) [reply internal/osbuildconfig/osbuildconfig_handler.go]: Changed it to OSBuildConfig with the name (which in the API description the field is called OSBuildConfig )

File: osbuildconfig_webhook_trigger.yaml (hunk 13 above)
Comment (bardielle) [reply osbuildconfig_webhook_trigger.yaml]: Theoretically, we can add trigger builds for osbuildconfigtemplates

File: main.go [hunk 16]
```diff
@@ -44,6 +50,10 @@ var (
 	setupLog = ctrl.Log.WithName("setup")
 )
 
+const (
+	Port = 8888
```
[the remark this answers is not in the document]
Comment (bardielle) [reply main.go:54]: It's an internal port (will be wrapped with a route)

File: internal/osbuildconfig/osbuildconfig_handler.go (hunk 14 above)
Comment (ygalblum) [reply internal/osbuildconfig/osbuildconfig_handler.go:67]: I checked back at GitHub and found that it does not allow adding custom fields into the payload of a webhook call. As a result, any information required by the receiver must be put into the URL. 
GitHub will send it's own payload as either JSON or URL Encoded.

Comment (bardielle) [reply internal/osbuildconfig/osbuildconfig_handler.go:68]: I changed it, now it's part of the header (openapi v3 has 4 types of parameters- path, query, header and cookie, more details can be found here  
https://swagger.io/docs/specification/describing-parameters/#header-parameters )
Please re-review @masayag @eloycoto @jakub-dzon

File: internal/osbuildconfig/osbuildconfig_handler.go [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+	return &OSBuildConfigHandler{
+		OSBuildConfigRepository: osBuildConfigRepository,
+		OSBuildRepository:       osBuildRepository,
+		SecretRepository:        secretRepository,
+		Scheme:                  scheme,
+		OSBuildCRCreator:        osBuildCRCreator,
+	}
+}
+func (o *OSBuildConfigHandler) TriggerBuild(w http.ResponseWriter, r *http.Request, namespace string, name string, params restapi.TriggerBuildParams) {
+	//func (o *OSBuildConfigHandler) TriggerBuild(w http.ResponseWriter, r *http.Request, namespace string, name string, secret string) {
```
Comment (jakub-dzon) [inline internal/osbuildconfig/osbuildconfig_handler.go:35]: Please remove commented-out code.

File: internal/repository/secret/secret.go [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+
+import (
+	"context"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+//go:generate mockgen -package=secret -destination=mock_secret.go . Repository
+type Repository interface {
+	Read(ctx context.Context, name string, namespace string) (*corev1.Secret, error)
```
[the remark this answers is not in the document]
Comment (jakub-dzon) [reply internal/repository/secret/secret.go:11]: [the role](https://github.com/project-flotta/osbuild-operator/blob/main/config/rbac/role.yaml)

File: main.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (bardielle) [reply main.go]: added a service - a router will be added in a different PR

File: internal/repository/secret/secret.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (bardielle) [reply internal/repository/secret/secret.go:11]: done

**Review by jakub-dzon — CHANGES_REQUESTED** [review]

**Comment by jakub-dzon** [comment]
> @jakub-dzon please re-review this PR, I rebaced over your merged code (osbuildconfigtemplate) and moved the `create` method to a common package, and tested it against a cluster (with several usecases). Still, since we don't have a UT on most of the code I'm afraid I break or change the functionality.

When the panic is fixed, the controller works as it should with templates.

File: config/httpapi/httpapi.yaml [hunk 19]
```diff
@@ -0,0 +1,71 @@
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: controller-httpapi
+  namespace: system
+  labels:
+    control-plane: controller-httpapi
```
[the remark this answers is not in the document]
Comment (bardielle) [reply config/httpapi/httpapi.yaml:7]: done

File: cmd/httpapi/main.go [hunk 20]
```diff
@@ -0,0 +1,129 @@
...
+	OSBuildConfigRepository := osbuildconfig.NewOSBuildConfigRepository(c)
+	SecretRepository := repositorysecret.NewSecretRepository(c)
+	OSBuildRepository := osbuild.NewOSBuildRepository(c)
+	OSBuildConfigTemplateRepository := osbuildconfigtemplate.NewOSBuildConfigTemplateRepository(c)
+	ConfigMapRepository := configmap.NewConfigMapRepository(c)
+	OSBuildCRCreator := manifests.NewOSBuildCRCreator()
+
+	h := restapi.Handler(internalosbuildconfig.NewOSBuildConfigHandler(OSBuildConfigRepository, OSBuildRepository, SecretRepository, scheme, OSBuildCRCreator, OSBuildConfigTemplateRepository, ConfigMapRepository))
+	server := &http.Server{
+		Addr:    fmt.Sprintf(":%v", Config.HttpsPort),
```
[the remark this answers is not in the document]
Comment (bardielle) [reply cmd/httpapi/main.go:74]: done

File: controllers/osbuildconfig_controller.go [hunk 21]
```diff
@@ -52,6 +43,7 @@ type OSBuildConfigReconciler struct {
 	OSBuildRepository               osbuild.Repository
 	OSBuildConfigTemplateRepository osbuildconfigtemplate.Repository
 	ConfigMapRepository             configmap.Repository
+	OSBuildCRCreator                manifests.OSBuildCRCreator
```
[the remark this answers is not in the document]
Comment (bardielle) [reply controllers/osbuildconfig_controller.go:46]: done

**Comment by bardielle** [comment]
@ygalblum and @jakub-dzon  can you approve/ add comments to that PR?

File: internal/httpapi/config.go [hunk 22]
```diff
@@ -0,0 +1,15 @@
+package httpapi
+
+type Config struct {
+	// The port of the HTTPs server
+	HttpPort uint16 `envconfig:"HTTPS_PORT" default:"8080"`
```
Comment (jakub-dzon) [inline internal/httpapi/config.go:5]: Please change the name of env variable to `HTTP_PORT`

File: internal/osbuildconfig/osbuildconfig_handler.go [hunk 23]
```diff
@@ -0,0 +1,96 @@
...
+	"k8s.io/apimachinery/pkg/api/errors"
+)
+
+var Config httpapi.Config
+
+type OSBuildConfigHandler struct {
+	OSBuildConfigRepository         repositoryosbuildconfig.Repository
+	SecretRepository                secret.Repository
+	OSBuildCRCreator                manifests.OSBuildCRCreator
+	OSBuildConfigTemplateRepository osbuildconfigtemplate.Repository
```
Comment (jakub-dzon) [inline internal/osbuildconfig/osbuildconfig_handler.go:25]: Please remove `OSBuildConfigTemplateRepository` and `ConfigMapRepository` - they are unused.

File: config/httpapi/kustomization.yaml [hunk 24]
```diff
@@ -0,0 +1,14 @@
...
+- httpapi.yaml
+
+generatorOptions:
+  disableNameSuffixHash: true
+
+configMapGenerator:
+- name: httpapi-config
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+images:
```
Comment (jakub-dzon) [inline config/httpapi/kustomization.yaml:11]: Please use a `quay.io/project-flotta`-located image or remove whole `images` section.

File: config/manager/kustomization.yaml [hunk 25]
```diff
@@ -5,6 +5,12 @@ generatorOptions:
...
 
 configMapGenerator:
-- name: manager-config
-  envs:
+- envs:
   - controller_manager_config.properties
+  name: manager-config
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+images:
```
Comment (jakub-dzon) [inline config/manager/kustomization.yaml:13]: Please use a `quay.io/project-flotta`-located image or remove `images`; maybe even revert the whole file.

File: internal/manifests/osbuild.go [hunk 26]
```diff
@@ -0,0 +1,215 @@
...
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+//go:generate mockgen -package=manifests -source=osbuild.go -destination=mock_osbuildcrcreator.go
+type OSBuildCRCreator interface {
+	Create(ctx context.Context, osBuildConfig *v1alpha1.OSBuildConfig) error
+}
+type OSBuildCreator struct {
+	OSBuildRepository               repositoryosbuild.Repository
+	Scheme                          *runtime.Scheme
```
Comment (jakub-dzon) [inline internal/manifests/osbuild.go:32]: Please group the members - repositories together, scheme before/after.

File: internal/httpapi/config.go [hunk 27]
```diff
@@ -0,0 +1,15 @@
...
+
+	// The port of the HTTPs server
+	ProbesPorts uint16 `envconfig:"PROBES_PORTS" default:"8081"`
```
Comment (ygalblum) [inline internal/httpapi/config.go:8]: Please rename to `ProbesPort` (as it is only one port)

File: internal/httpapi/config.go [hunk 28]
```diff
@@ -0,0 +1,15 @@
+package httpapi
+
+type Config struct {
```
Comment (ygalblum) [inline internal/httpapi/config.go:3]: Consider moving the loading of the configuration to this file (to encapsulate the implementation). Also saving it to a global variable will allow fetching it from everywhere without having to keep readying the env

File: internal/logger/logger.go [hunk 29]
```diff
@@ -0,0 +1,21 @@
+package logger
+
+import (
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+func Logger(logLevel string) (error, *zap.SugaredLogger) {
```
Comment (ygalblum) [inline internal/logger/logger.go:8]: AFAIK `error` is usually the last returned values, no?

File: internal/manifests/osbuild.go [hunk 30]
```diff
@@ -0,0 +1,215 @@
...
+	repositoryosbuildconfig "github.com/project-flotta/osbuild-operator/internal/repository/osbuildconfig"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+//go:generate mockgen -package=manifests -source=osbuild.go -destination=mock_osbuildcrcreator.go
+type OSBuildCRCreator interface {
+	Create(ctx context.Context, osBuildConfig *v1alpha1.OSBuildConfig) error
+}
```
Comment (ygalblum) [inline internal/manifests/osbuild.go:29]: Please add an empty line

File: internal/manifests/osbuild.go [hunk 31]
```diff
@@ -0,0 +1,215 @@
...
+	"github.com/project-flotta/osbuild-operator/internal/repository/osbuildconfigtemplate"
+	"github.com/project-flotta/osbuild-operator/internal/templates"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	v1alpha1 "github.com/project-flotta/osbuild-operator/api/v1alpha1"
```
Comment (ygalblum) [inline internal/manifests/osbuild.go:18]: Please give this import a meaningful name

File: internal/osbuildconfig/osbuildconfig_handler.go [hunk 32]
```diff
@@ -0,0 +1,92 @@
...
+
+type OSBuildConfigHandler struct {
+	OSBuildConfigRepository repositoryosbuildconfig.Repository
+	SecretRepository        secret.Repository
+	OSBuildCRCreator        manifests.OSBuildCRCreator
+}
+
+func NewOSBuildConfigHandler(osBuildConfigRepository repositoryosbuildconfig.Repository,
+	secretRepository secret.Repository, osBuildCRCreator manifests.OSBuildCRCreator) *OSBuildConfigHandler {
+	err := envconfig.Process("", &Config)
```
Comment (ygalblum) [inline internal/osbuildconfig/osbuildconfig_handler.go:27]: Why not get this either at the API or from a global variable?

File: osbuildconfig_webhook_trigger.yaml [hunk 33]
```diff
@@ -0,0 +1,76 @@
...
+        "401":
+          description: Unauthorized
+        "403":
+          description: Forbidden
+        "404":
+          description: Error
+        "500":
+          description: Error
+components:
+  securitySchemes:
```
Comment (ygalblum) [inline osbuildconfig_webhook_trigger.yaml:62]: Since we don't use any securityScheme, this block can be removed

**Review by ygalblum — COMMENTED** [review]
I commented in some places about the imports, but I think there are more
1. Please order the imports - generic golang, external, internal
2. Please give the imports meaningful names

File: cmd/httpapi/main.go [hunk 34]
```diff
@@ -0,0 +1,129 @@
+package main
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/kelseyhightower/envconfig"
+	"github.com/project-flotta/osbuild-operator/api/v1alpha1"
```
[the remark this answers is not in the document]
Comment (bardielle) [reply cmd/httpapi/main.go:8]: In a different comment, I was asked to change it to `v1aplha1` 
I will change it to `osbuildv1alpha1`

File: Makefile [hunk 35]
```diff
@@ -222,8 +234,9 @@ install-cert-manager: cmctl
 	${CMCTL} check api --wait=5m
 
 .PHONY: deploy
-deploy: manifests kustomize install-cert-manager ## Deploy controller to the K8s cluster specified in ~/.kube/config.
+deploy: manifests kustomize  ## Deploy controller to the K8s cluster specified in ~/.kube/config.
```
[the remark this answers is not in the document]
Comment (bardielle) [reply Makefile:237]: by a mistake - I still have a bug around this area in my cluster
After deploying the operator and trying to apply the deployment again, the install-cert-manager is stuck forever that's why I had to remove it for my local env, but I didn't mean to push it

**Review by jakub-dzon — APPROVED** [review]

File: Makefile [hunk 36]
```diff
@@ -121,11 +121,22 @@ ifeq (, $(shell which mockery))
 endif
 ifeq (, $(shell which mockgen))
 	(cd /tmp/ && go install github.com/golang/mock/mockgen@v1.6.0)
+endif
+ifeq (, $(shell which oapi-codegen))
+	(cd /tmp/ && go install github.com/deepmap/oapi-codegen/cmd/oapi-codegen@latest)
 endif
 	@exit
 
+.PHONY: generate-oapi
```
[the remark this answers is not in the document]
Comment (bardielle) [reply Makefile]: Done, it's part of the file `internal/osbuildconfig/osbuildconfig_handler.go`

File: internal/osbuildconfig/osbuildconfig_handler.go [hunk 37]
```diff
@@ -0,0 +1,95 @@
...
+
+type OSBuildConfigHandler struct {
+	OSBuildConfigRepository repositoryosbuildconfig.Repository
+	SecretRepository        secret.Repository
+	OSBuildCRCreator        manifests.OSBuildCRCreator
+}
+
+func NewOSBuildConfigHandler(osBuildConfigRepository repositoryosbuildconfig.Repository,
+	secretRepository secret.Repository, osBuildCRCreator manifests.OSBuildCRCreator) *OSBuildConfigHandler {
+	err := httpapi.Load()
```
Comment (ygalblum) [inline internal/osbuildconfig/osbuildconfig_handler.go:30]: Why do you need to load the configuration here? it was already loaded in main

Comment (bardielle) [reply internal/osbuildconfig/osbuildconfig_handler.go]: removed it

**Review by ygalblum — APPROVED** [review]
